### PR TITLE
Create a USSO login link component

### PR DESCRIPTION
**Pull request by hatched**
Created a USSO login link component and use it in the canvas and the deployment flow.

### CODE REVIEW

**Comment by hatched** [comment]
#### To QA

Try and log in using Juju 1, Juju 2, Anon login link on the top right, Anon DF

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6591/

File: jujugui/static/gui/src/app/app.js [hunk 1]
```diff
@@ -989,32 +990,44 @@ YUI.add('juju-gui', function(Y) {
...
         // Do not show the logout link if the user is in sandbox mode.
         return;
       }
-      // If the charmbrowser is open then don't show the logout link.
-      var visible = !this.state.current.store;
-      var charmstore = this.get('charmstore');
+      const controllerAPI = this.controllerAPI;
+      const linkContainer = document.getElementById('profile-link-container');
       const bakeryFactory = this.bakeryFactory;
+      if (controllerAPI && !controllerAPI.userIsAuthenticated) {
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/app.js]: Should `linkContainer` be null checked here?

File: jujugui/static/gui/src/app/components/usso-login-link/usso-login-link.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    */
+    _handleLogin: function(e) {
+      if (e && e.preventDefault) {
+        // Depending on the login link type there may or may not be a
+        // preventDefault method.
+        e.preventDefault();
+      }
+      this.props.loginToController(err => {
+        if (err) {
+          console.error('cannot log into the controller:', err);
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/usso-login-link/usso-login-link.js]: Can you sentence case this log please.

File: jujugui/static/gui/src/app/components/usso-login-link/usso-login-link.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+    /**
+      Returns the text login link.
+    */
+    _renderTextLink: function() {
+      return (
+        <a className={'logout-link'}
+          onClick={this._handleLogin}
+          target="_blank">
+          Login
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/usso-login-link/usso-login-link.js]: Shouldn't this link also say `Sign up our login`?

File: jujugui/static/gui/src/app/components/usso-login-link/usso-login-link.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+          action={this._handleLogin}
+          type="positive"
+          title="Sign up or Login" /> );
+    },
+
+    render: function() {
+      let linkFn = this._renderTextLink;
+      if (this.props.displayType === 'button') {
+        linkFn = this._renderButtonLink;
+      }
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/usso-login-link/usso-login-link.js]: This is creating a component and then overriding the component if button is requested. This is not very efficient.

File: jujugui/static/gui/src/app/components/usso-login-link/test-usso-login-link.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
+/*
+This file is part of the Juju GUI, which lets users view and manage Juju
+environments within a graphical interface (https://launchpad.net/juju-gui).
+Copyright (C) 2015 Canonical Ltd.
```
Comment (frankban) [inline jujugui/static/gui/src/app/components/usso-login-link/test-usso-login-link.js]: 2017

File: jujugui/static/gui/src/app/app.js [hunk 6]
```diff
@@ -989,32 +990,44 @@ YUI.add('juju-gui', function(Y) {
     },
 
     /**
-      Renders the Logout component.
-
-      @method _renderLogout
+      Renders the Logout component or log in link depending on the environment
```
Comment (frankban) [inline jujugui/static/gui/src/app/app.js]: s/Logout/log out/

File: jujugui/static/gui/src/app/components/usso-login-link/usso-login-link.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
+/*
+This file is part of the Juju GUI, which lets users view and manage Juju
+environments within a graphical interface (https://launchpad.net/juju-gui).
+Copyright (C) 2015 Canonical Ltd.
```
Comment (frankban) [inline jujugui/static/gui/src/app/components/usso-login-link/usso-login-link.js]: 2017

File: jujugui/static/gui/src/app/components/usso-login-link/usso-login-link.js [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    _renderButtonLink: function() {
+      return (
+        <juju.components.GenericButton
+          action={this._handleLogin}
+          type="positive"
+          title="Sign up or Login" /> );
+    },
+
+    render: function() {
+      let linkFn = this._renderTextLink;
```
Comment (frankban) [inline jujugui/static/gui/src/app/components/usso-login-link/usso-login-link.js]: Or just:
```javascript
render: function() {
  if (this.props.displayType === 'button') {
    return this._renderButtonLink();
  }
  return this._renderTextLink();
}
```
?

File: jujugui/static/gui/src/app/components/usso-login-link/usso-login-link.js (hunk 4 above)
Comment (frankban) [inline jujugui/static/gui/src/app/components/usso-login-link/usso-login-link.js]: ISTM that this is just storing the reference to the widget in a separate var, not instantiating the widgets. By the way, I also suggested a slightly different way to do it, which IMHO requires less thinking.

File: jujugui/static/gui/src/app/components/usso-login-link/usso-login-link.js (hunk 3 above)
Comment (frankban) [inline jujugui/static/gui/src/app/components/usso-login-link/usso-login-link.js]: In that context just Login is good enough IMHO.

File: jujugui/static/gui/src/app/components/usso-login-link/usso-login-link.js (hunk 2 above)
Comment (frankban) [inline jujugui/static/gui/src/app/components/usso-login-link/usso-login-link.js]: Lower case errors actually allow them to be more easily concatenated. [shrug]

**Comment by frankban** [comment]
QA problem: when clicking to log in (both canvas and deployment flow), nothing happens and I am getting `cannot log into the controller: authentication failed: invalid entity name or password`

File: jujugui/static/gui/src/app/app.js (hunk 1 above)
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: done

File: jujugui/static/gui/src/app/components/usso-login-link/test-usso-login-link.js (hunk 5 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/usso-login-link/test-usso-login-link.js]: done

File: jujugui/static/gui/src/app/app.js (hunk 6 above)
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: done

File: jujugui/static/gui/src/app/components/usso-login-link/usso-login-link.js (hunk 7 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/usso-login-link/usso-login-link.js]: done

File: jujugui/static/gui/src/app/components/usso-login-link/usso-login-link.js (hunk 8 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/usso-login-link/usso-login-link.js]: done

**Comment by frankban** [comment]
Nevermind, QA is actually ok. Thanks!

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6593/

**Comment by anthonydillon** [comment]
Thanks @hatched, Code :+1:

**Comment by hatched** [comment]
Thanks all! :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**